source/material_model/composition_reaction.cc

```diff
@@ -63,16 +64,18 @@ namespace aspect
                 break;
             }
 
-          const double c1 = composition.size()>0?
-                            std::max(0.0, composition[0])
-                            :
-                            0.0;
-          const double c2 = composition.size()>1?
-                            std::max(0.0, composition[1])
-                            :
-                            0.0;
-          out.densities[i] = reference_rho * (1 - thermal_alpha * (temperature - reference_T))
-                             + compositional_delta_rho_1 * c1 + compositional_delta_rho_2 * c2;
+          // The Composition reaction model has up to two compositional field that can influence the density
+          EquationOfStateOutputs<dim> eos_outputs (in.composition[i].size()>2 ? 3 : in.composition[i].size() + 1);
```

Comment:
Creating this object is somewhat slow. Can we move it outside of the loop over `i`? Same for the other material models.